### ISSUE TITLE
High memory usage on Kubernetes (~400MB with close to zero traffic)

### ISSUE THREAD

**Issue by christophetd** [linkerd/linkerd#1689]
Might be related to #1343. 

I'm running two linkerd instances: one as a sidecar instance, and the other one as an edge proxy.

```
            +-----------------------------------+                  +---------------------------------+
            |                                   |                  |                                 |
            |                                   |                  |                                 |
            |                                   |                  |  +-----------+       +--------+ |
            |          LinkerD Edge             |                  |  |           |       |        | |
+----------->                                   +---------------------> LinkerD   +------->  App   | |
            |                                   |                  |  +-----------+       +--------+ |
            |                                   |                  |                                 |
            |                                   |                  |                                 |
            +-----------------------------------+                  +---------------------------------+
``` 

Those two instance seem to take way too much memory, even there is little to no traffic. I made a test where I hit the linkerd edge instance (itself hitting the linkerd sidecar instance) at a rate of 2 requests per second, and I'm getting a memory usage of 300-400 MB (which obviously isn't viable):

![image](https://user-images.githubusercontent.com/136675/32223438-a467ee02-be3d-11e7-972e-09fd56a07e01.png)
*LinkerD sidecar instance memory usage*

![image](https://user-images.githubusercontent.com/136675/32223492-d320f75c-be3d-11e7-914a-7b8c20f0d1a6.png)
*LinkerD edge instance memory usage*

I'm not sure this is normal, but according to some other benchmarks I've seen, I think the memory footprint of LinkerD is supposed to be much lower. Any idea of what could be the cause?

Additional details:
- **Environment**: OpenShift v1.5.1 running Kubernetes v1.5.2
- **LinkerD** version: 1.3.0
- **Configuration files**: see [here](https://gist.github.com/christophetd/20a78bb137e2493431c287c018fe30b4)
- **JVM stats from /admin/metrics.json**: see [here](https://gist.github.com/christophetd/3b75d9df54ad6ad97d94a7a95e938a1d)

The Kubernetes cluster I'm running LinkerD on will be upgraded to a more recent Kubernetes version shortly, I'll let you know if that changes anything.

Thanks!

**Comment by christophetd** [comment]
Note that the memory usage doesn't seem to grow bigger if I hit LinkerD with a higher request rate (I tested 10-20rps).

**Comment by siggy** [comment]
Hi @christophetd, thanks for the detailed report. We're are currently investigating possibly related memory usage/leak issues, follow along at #1685 and #1690. Stay tuned.

**Comment by siggy** [comment]
Quick update for folks watching this issue. We have reproduced a leak and are actively working on a fix. To confirm the leak you are seeing is the same one we have identified, have a look at your `open_streams` metrics. If they grow over time, that is the leak.

**Comment by siggy** [comment]
Hi @christophetd, following up on this. Can you confirm whether you are seeing memory usage grow unbounded over time, or just high memory usage in general?

A few things to try:
- Set `JVM_HEAP_MIN` and `JVM_HEAP_MAX` environment variables to something like `256M` or `512M`. It's important to set them to the same value, to minimize memory fragmentation.
- zipkin `sampleRate: 1` is very resource intensive. Do you observe any difference in memory usage if you set this to `0.1` or `0.01`?

**Comment by christophetd** [comment]
It seems to me that the memory usage is just high in general, but doesn't particularly grow unbounded. I will try some of the remediations suggested and let you know if it lowers the memory footprint of Linkerd.

**Comment by hawkw** [comment]
Hi @christophetd, we suspect the leak you're seeing may be the same as #1694, which was fixed by @sgrankin in #1714. We've just published a release candidate Docker tag `buoyantio/linkerd:1.3.3-RC1-jdk`; can we get you to try using that image and confirm that this issue is fixed?

**Comment by wmorgan** [comment]
@christophetd I'm going to close this issue, since I believe we've fixed the underlying problem. If you still see this on 1.3.7 or beyond, please feel free to reopen.
